### PR TITLE
Introduce Auth Metadata and Identity Grpc Service and support Pkce auth in admin client

### PR DESCRIPTION
**Pull request by EngHabu**
Signed-off-by: Haytham Abuelfutuh <haytham@afutuh.com>

# TL;DR
* Add AuthMetadata and Identity Grpc Service definitions
* Update admin client to:
  * Simplify config to not require anything besides admin's endpoint,
  * Support Pkce and ClientSecret auth flows in the client,
  * Build an admin Clientset (ala k8s Clientsets) to expose all admin clients (with shared connection).
  * Support an external TokenCache (e.g. Keyring-based) as an optional component.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/925

### CODE REVIEW

File: clients/go/admin/token_source.go [hunk 1]
```diff
@@ -27,10 +27,11 @@ func (ts CustomHeaderTokenSource) GetRequestMetadata(ctx context.Context, uri ..
...
 }
 
-// Even though Admin is capable of serving authentication without SSL, we're going to require it here. That is, this module's
-// canonical Admin client will only do auth over SSL.
+// RequireTransportSecurity returns whether this credentials class requires TLS/SSL. OAuth uses Bearer tokens that are
+// susceptible to MITM (Man-In-The-Middle) attacks that are mitigated by TLS/SSL. We return false here to make it easier
+// to setup auth. However, in a production environment, TLS for Auth is a requirement.
 func (ts CustomHeaderTokenSource) RequireTransportSecurity() bool {
-	return true
+	return false
```
Comment (pmahindrakar-oss) [inline clients/go/admin/token_source.go:34]: Shouldn't this be configurable, trying to see how do we turn this on for production usecases as there checks in oauth that require this to be true on a secure channel.
eg : this check in credentials.go of grpc
```
credentials.CheckSecurityLevel(ri.AuthInfo, credentials.PrivacyAndIntegrity)
```

File: clients/go/admin/client.go [hunk 2]
```diff
@@ -54,105 +78,153 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
...
+
+		return []grpc.DialOption{opt}, nil
+	}
+
+	return nil, nil
+}
+
+func NewAdminConnection(_ context.Context, cfg *Config, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	if opts == nil {
+		opts = make([]grpc.DialOption, 0, 10)
```
Comment (pmahindrakar-oss) [inline clients/go/admin/client.go:153]: nit : May be somewhere we can mention the reasoning for this size which can act as a reference for required set of opts for authorization versus normal ones when talking to admin

**Review by pmahindrakar-oss — APPROVED** [review]

File: clients/go/admin/token_source.go (hunk 1 above)
Comment (EngHabu) [reply clients/go/admin/token_source.go]: done

File: clients/go/admin/client.go (hunk 2 above)
Comment (EngHabu) [reply clients/go/admin/client.go]: done

File: clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package interfaces
```
Comment (kumare3) [inline clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go:1]: the file name is wrong, 
orchest`ar`tor

File: clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+
+	"golang.org/x/oauth2"
+)
+
+//go:generate mockery -all -case=underscore
+
+type FetchTokenOrchestrator interface {
+	RefreshTheToken(ctx context.Context, clientConf *oauth2.Config, token *oauth2.Token) (*oauth2.Token, error)
```
Comment (kumare3) [inline clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go:14]:
```suggestion
	RefreshToken(ctx context.Context, clientConf *oauth2.Config, token *oauth2.Token) (*oauth2.Token, error)
```

File: clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	FetchTokenFromCacheOrRefreshIt(ctx context.Context) (*oauth2.Token, error)
```
Comment (kumare3) [inline clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go:15]: docs?

File: clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+	"context"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+
+	"golang.org/x/oauth2"
+)
+
+//go:generate mockery -all -case=underscore
+
+type FetchTokenOrchestrator interface {
```
Comment (kumare3) [inline clients/go/admin/threelegauth/interfaces/fetch_token_orchestartor.go:13]: docs?

File: clients/go/admin/threelegauth/token_cache_with_keyring.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
+package threelegauth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/zalando/go-keyring"
```
Comment (kumare3) [inline clients/go/admin/threelegauth/token_cache_with_keyring.go:10]: should we introduce this dependency in flyteidl? Seems a very client specific thing?

File: clients/go/admin/threelegauth/token_cache_with_keyring.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+	"golang.org/x/oauth2"
+)
+
+var defaultCacheProvider = TokenCacheProvider{
```
Comment (kumare3) [inline clients/go/admin/threelegauth/token_cache_with_keyring.go:14]: docs for this? Maybe this struct level docs will help to understand that this uses a Keyring based token cache

File: clients/go/admin/threelegauth/token_cache_with_keyring.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	serviceUser: "flytectl-user",
+	serviceName: "flytectl",
+}
+
+type TokenCacheProvider struct {
```
Comment (kumare3) [inline clients/go/admin/threelegauth/token_cache_with_keyring.go:19]:
```suggestion
type KeyringTokenCacheProvider? struct {
```

File: clients/go/admin/threelegauth/oauth2_client.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	pkceCodeVerifier = ""
+	return codeVerifier
+}
+
+func state(n int) string {
+	data := randomString(n)
+	return base64.RawURLEncoding.EncodeToString([]byte(data))
+}
+
+type DefaultHeaderTokenSource struct {
```
Comment (kumare3) [inline clients/go/admin/threelegauth/oauth2_client.go:85]: docs on this public structure

File: clients/go/admin/threelegauth/auth_flow_orchestrator.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/pkg/browser"
```
Comment (kumare3) [inline clients/go/admin/threelegauth/auth_flow_orchestrator.go:13]: wow

File: clients/go/admin/threelegauth/auth_flow_orchestrator.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+	pkceCodeChallenge = generateCodeChallenge(pkceCodeVerifier)
+	stateString = state(32)
+	nonces = state(32)
+	// Replace S256 with one from cient config and provide a support to generate code challenge using the passed
+	// in method.
+	urlToOpen := clientConf.AuthCodeURL(stateString) + "&nonce=" + nonces + "&code_challenge=" +
+		pkceCodeChallenge + "&code_challenge_method=S256"
+
+	go func() {
+		if err = http.ListenAndServe(redirectURL.Host, nil); err != nil {
```
Comment (kumare3) [inline clients/go/admin/threelegauth/auth_flow_orchestrator.go:97]: how do we shut this server down?

**Comment by kumare3** [comment]
Some minor comments, looks good to me otherwise

**Review by kumare3 — APPROVED** [review]

File: protos/flyteidl/service/auth.proto [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    string issuer = 1;
+
+    // URL of the authorization server's authorization endpoint [RFC6749]. This is REQUIRED unless no grant types are
+    // supported that use the authorization endpoint.
+    string authorization_endpoint = 2;
+
+    // URL of the authorization server's token endpoint [RFC6749].
+    string token_endpoint = 3;
+
+    // JSON array containing a list of the OAuth 2.0 "response_type" values that this authorization server supports.
```
Comment (katrogan) [inline protos/flyteidl/service/auth.proto:38]: nit: if this is a repeated field, it's not JSON?

File: clients/go/admin/pkce/token_cache_inmemory.go [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+	"fmt"
+
+	"golang.org/x/oauth2"
+)
+
+type TokenCacheInMemoryProvider struct {
+	token *oauth2.Token
+}
+
+func (t *TokenCacheInMemoryProvider) SaveToken(token *oauth2.Token) error {
```
Comment (katrogan) [inline clients/go/admin/pkce/token_cache_inmemory.go:13]: if this is the only impl, why even have an error as a response?

File: clients/go/admin/pkce/token_cache.go [hunk 15]
```diff
@@ -0,0 +1,14 @@
+package pkce
+
+import "golang.org/x/oauth2"
+
+//go:generate mockery -all -case=underscore
+
+// TokenCache defines the interface needed to cache and retrieve oauth tokens.
```
Comment (katrogan) [inline clients/go/admin/pkce/token_cache.go:7]: why is this even an interface? ease of testing?

File: clients/go/admin/pkce/oauth2_client.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+// challenge.
+func generateCodeChallenge(codeVerifier string) string {
+	// Create a sha-265 hash from the code verifier...
+	s256 := sha256.New()
+	_, _ = s256.Write([]byte(codeVerifier))
+	// Then base64 encode the hash sum to create a code challenge...
+	return base64.RawURLEncoding.EncodeToString(s256.Sum(nil))
+}
+
+func state(n int) string {
```
Comment (katrogan) [inline clients/go/admin/pkce/oauth2_client.go:46]: what does state mean here?

File: clients/go/admin/pkce/auth_flow_orchestrator.go [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+	ts := f.clientConfig.TokenSource(ctx, token)
+	var refreshedToken *oauth2.Token
+	var err error
+	refreshedToken, err = ts.Token()
+	if err != nil {
+		logger.Warnf(ctx, "failed to refresh the token due to %v and will be doing re-auth", err)
+		return nil, err
+	}
+
+	if refreshedToken != nil {
```
Comment (katrogan) [inline clients/go/admin/pkce/auth_flow_orchestrator.go:40]: can err != nil but refreshToken ever == nil ?

File: clients/go/admin/pkce/auth_flow_orchestrator.go [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"github.com/flyteorg/flytestdlib/logger"
+
+	"github.com/pkg/browser"
+	"golang.org/x/oauth2"
+)
+
+const (
+	Timeout     = 15 * time.Second
```
Comment (katrogan) [inline clients/go/admin/pkce/auth_flow_orchestrator.go:18]: should this be configurable?

File: clients/go/admin/pkce/auth_flow_orchestrator.go [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+				logger.Errorf(ctx, "unable to save the new token due to %v", err)
+				return nil, err
+			}
+		}
+	}
+
+	return refreshedToken, nil
+}
+
+// FetchTokenFromCacheOrRefreshIt Fetch token from cache or refresh it. ignore the error while fetching from cache
```
Comment (katrogan) [inline clients/go/admin/pkce/auth_flow_orchestrator.go:54]: nit: not true? you return the error in line 58 below

File: clients/go/admin/pkce/auth_flow_orchestrator.go [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+	if token.Expiry.Add(RefreshTime).Before(time.Now()) {
+		return token, nil
+	}
+
+	token, err = f.RefreshToken(ctx, token)
+	if err != nil {
+		return nil, fmt.Errorf("failed to refresh token using cached token. Error: %w", err)
+	}
+
+	if !token.Valid() {
```
Comment (katrogan) [inline clients/go/admin/pkce/auth_flow_orchestrator.go:75]: what if token is nil?

File: clients/go/admin/client.go [hunk 21]
```diff
@@ -54,105 +81,195 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
...
+	if err != nil {
+		logger.Panicf(ctx, "Failed to initialize client. Error: %v", err)
+		return nil
+	}
+
+	return set.AdminClient()
+}
+
+// initializeClients creates an AdminClient and AuthServiceClient with a shared Admin connection for the process.
+// Note that if called with different cfg/dialoptions, it will not
```
Comment (katrogan) [inline clients/go/admin/client.go:221]: will not?

File: clients/go/admin/client.go [hunk 22]
```diff
@@ -54,105 +81,195 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
...
+		authMetadataClient, err := InitializeAuthMetadataClient(ctx, cfg)
+		if err != nil {
+			logger.Panicf(ctx, "failed to initialize Auth Metadata Client. Error: %v", err)
+		}
+
+		// If auth is enabled, this call will return the required information to use to authenticate, otherwise,
+		// start the client without authentication.
+		opt, err := getAuthenticationDialOption(ctx, cfg, tokenCache, authMetadataClient)
+		if err != nil {
+			logger.Warnf(ctx, "Starting an unauthenticated client because: %v", err)
```
Comment (katrogan) [inline clients/go/admin/client.go:233]: we always warn then if auth is not enabled?

**Review by katrogan — APPROVED** [review]
sorry, a lot of questions

File: clients/go/admin/pkce/token_cache_inmemory.go (hunk 14 above)
Comment (EngHabu) [reply clients/go/admin/pkce/token_cache_inmemory.go:13]: The other implementation is in flytectl... it's pluggable when building the Clientset (see NewClientSet().WithTokenCache)

File: clients/go/admin/pkce/token_cache.go (hunk 15 above)
Comment (EngHabu) [reply clients/go/admin/pkce/token_cache.go:7]: We actually want another implementation (keyring based like the one you added to flyte-cli) but didn't want to put it here with the dependencies it brings on... so left it in flytectl...

File: clients/go/admin/pkce/oauth2_client.go (hunk 16 above)
Comment (EngHabu) [reply clients/go/admin/pkce/oauth2_client.go:46]: Defined in the spec here: https://tools.ietf.org/html/rfc6749#section-4.1.1
it's used to prevent XSS attacks...

File: clients/go/admin/pkce/auth_flow_orchestrator.go (hunk 17 above)
Comment (EngHabu) [reply clients/go/admin/pkce/auth_flow_orchestrator.go:40]: yes.. refreshToken is optional..

File: clients/go/admin/pkce/auth_flow_orchestrator.go (hunk 18 above)
Comment (EngHabu) [reply clients/go/admin/pkce/auth_flow_orchestrator.go:18]: I was on the fence... alright, will move it to config

File: clients/go/admin/pkce/auth_flow_orchestrator.go (hunk 20 above)
Comment (EngHabu) [reply clients/go/admin/pkce/auth_flow_orchestrator.go:75]: Valid() checks for that... nifty (or hacky?) go feature..

File: clients/go/admin/client.go (hunk 21 above)
Comment (EngHabu) [reply clients/go/admin/client.go:221]: oops.. thanks

File: clients/go/admin/client.go (hunk 22 above)
Comment (EngHabu) [reply clients/go/admin/client.go:233]: yeah... an incentive to enable auth? thoughts?

File: clients/go/admin/pkce/token_cache_inmemory.go (hunk 14 above)
Comment (katrogan) [reply clients/go/admin/pkce/token_cache_inmemory.go:13]: oh, thanks for explaining!

File: clients/go/admin/pkce/token_cache.go (hunk 15 above)
Comment (katrogan) [reply clients/go/admin/pkce/token_cache.go:7]: thanks for explaining!

File: clients/go/admin/pkce/auth_flow_orchestrator.go (hunk 20 above)
Comment (katrogan) [reply clients/go/admin/pkce/auth_flow_orchestrator.go:75]: oh! TIL

File: clients/go/admin/client.go (hunk 22 above)
Comment (katrogan) [reply clients/go/admin/client.go:233]: as long as it's intentional, sounds good :)

**Review by kumare3 — APPROVED** [review]

**Review by katrogan — APPROVED** [review]
